### PR TITLE
internal/blobstore: check content hash when uploading

### PR DESCRIPTION
**Pull request by rogpeppe**
We can now check the content hash with the new method implemented in juju/blobstore.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/86/

File: internal/blobstore/blobstore_test.go [hunk 1]
```diff
@@ -54,6 +54,46 @@ func (s *BlobStoreSuite) TestPutOpen(c *gc.C) {
...
+	c.Assert(err, gc.ErrorMatches, "hash mismatch")
+	c.Assert(chal, gc.IsNil)
+
+	rc, length, err := store.Open(hashOf(content))
+	c.Assert(err, gc.ErrorMatches, "resource.*not found")
+	c.Assert(rc, gc.Equals, nil)
+	c.Assert(length, gc.Equals, int64(0))
+}
+
+func (s *BlobStoreSuite) TestPutUnchallenged(c *gc.C) {
```
Comment (frankban) [inline internal/blobstore/blobstore_test.go:70]: Thanks for adding this missing test.

**Comment by frankban** [comment]
:+1: Thanks for the branch Roger.
What do you think about adding a wrong hash test also for the POST API endpoint?
The underlying blobstore usage is already tested, but it would be nice to have a more end-to-end check.

**Comment by jrwren** [comment]
:+1:

**Comment by rogpeppe** [comment]
yes. i'm adding that in a followup branch.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**